File: solutions/RefactoredWeatherSwift/LegacyWeatherSwiftTests/NetworkDependencyTests.swift
//
//  NetworkDependencyTests.swift
//  
//  DEMONSTRATES: Exercise 4 - Network Dependency Breaking Testing
//  
//  Shows how network dependency abstraction enables fast, reliable testing
//  of async network operations without real HTTP calls
//

import XCTest
@testable import LegacyWeatherSwift

// MARK: - Mock Network Service for Testing
// Mock implementation for testing without real network calls
class MockNetworkService: WeatherNetworkService {
    
    // Test control properties
    var mockData: Data?
    var mockError: Error?
    var requestDelay: TimeInterval = 0
    var capturedURLs: [URL] = []
    
    // Configurable responses for different scenarios
    enum NetworkScenario {
        case success(Data)
        case failure(Error)
        case timeout
        case malformedData
    }
    
    private var scenario: NetworkScenario = .success(Data())
    
    func configure(scenario: NetworkScenario) {
        self.scenario = scenario
    }
    
    func fetchWeatherData(from url: URL) async throws -> Data {
        // Capture the URL for verification in tests
        capturedURLs.append(url)
        
        // Simulate network delay if configured
        if requestDelay > 0 {
            try await Task.sleep(nanoseconds: UInt64(requestDelay * 1_000_000_000))
        }
        
        // Return response based on configured scenario
        switch scenario {
        case .success(let data):
            return data
        case .failure(let error):
            throw error
        case .timeout:
            throw URLError(.timedOut)
        case .malformedData:
            throw URLError(.cannotParseResponse)
        }
    }
}

// MARK: - Network Error Types for Testing
enum TestNetworkError: Error, Equatable {
    case noConnection
    case serverError
    case invalidResponse
    
    var localizedDescription: String {
        switch self {
        case .noConnection: return "No network connection"
        case .serverError: return "Server error occurred"
        case .invalidResponse: return "Invalid server response"
        }
    }
}

// Test-friendly WeatherSingleton for network dependency tests
class NetworkTestWeatherSingleton: WeatherSingleton {
    override func loadInitialData() {
        // Override to prevent automatic network calls during test initialization
    }
}

// MARK: - Tests
class NetworkDependencyTests: XCTestCase {
    
    var weatherManager: NetworkTestWeatherSingleton!
    var mockNetworkService: MockNetworkService!
    
    override func setUp() {
        super.setUp()
        mockNetworkService = MockNetworkService()
        weatherManager = NetworkTestWeatherSingleton(
            timeProvider: SystemTimeProvider(),
            networkService: mockNetworkService
        )
    }
    
    override func tearDown() {
        weatherManager = nil
        mockNetworkService = nil
        super.tearDown()
    }
    
    // MARK: - Successful Network Request Tests
    
    func test_fetchWeather_success_parsesDataCorrectly() async {
        // ARRANGE: Valid weather JSON response
        let validWeatherJSON = """
        {
            "name": "London",
            "main": {
                "temp": 22.5
            },
            "weather": [
                {
                    "description": "clear sky"
                }
            ]
        }
        """.data(using: .utf8)!
        
        mockNetworkService.configure(scenario: .success(validWeatherJSON))
        
        // ACT: Trigger weather fetch
        weatherManager.fetchWeather(for: "London")
        
        // Wait a moment for async operation
        try? await Task.sleep(nanoseconds: 100_000_000) // 0.1 second
        
        // ASSERT: Weather data should be populated
        XCTAssertNotNil(weatherManager.currentWeather, "Weather data should be set after successful network request")
        XCTAssertEqual(weatherManager.currentWeather?.cityName, "London")
        if let temperature = weatherManager.currentWeather?.temperature {
            XCTAssertEqual(temperature, 22.5, accuracy: 0.1)
        } else {
            XCTFail("Temperature should not be nil")
        }
        XCTAssertEqual(weatherManager.currentWeather?.description, "clear sky")
        XCTAssertFalse(weatherManager.isLoading, "Loading should be false after completion")
        XCTAssertTrue(weatherManager.errorMessage.isEmpty, "Error message should be empty on success")
    }
    
    func test_fetchWeather_verifies_correct_URL_construction() async {
        // ARRANGE: Mock successful response
        mockNetworkService.configure(scenario: .success(Data()))
        
        // ACT: Fetch weather for specific city
        weatherManager.fetchWeather(for: "Tokyo")
        
        // Wait for async operation
        try? await Task.sleep(nanoseconds: 50_000_000) // 0.05 second
        
        // ASSERT: Verify correct URL was requested
        XCTAssertEqual(mockNetworkService.capturedURLs.count, 1, "Should make exactly one network request")
        
        let capturedURL = mockNetworkService.capturedURLs.first!
        XCTAssertTrue(capturedURL.absoluteString.contains("Tokyo"), "URL should contain city name 'Tokyo'")
        XCTAssertTrue(capturedURL.absoluteString.contains("api.openweathermap.org"), "URL should use correct API endpoint")
        XCTAssertTrue(capturedURL.absoluteString.contains("units=metric"), "URL should request metric units")
    }
    
    // MARK: - Network Error Handling Tests
    
    func test_fetchWeather_networkError_handlesGracefully() async {
        // ARRANGE: Configure network error
        let networkError = TestNetworkError.noConnection
        mockNetworkService.configure(scenario: .failure(networkError))
        
        // ACT: Attempt to fetch weather
        weatherManager.fetchWeather(for: "Paris")
        
        // Wait for async operation
        try? await Task.sleep(nanoseconds: 100_000_000)
        
        // ASSERT: Error should be handled gracefully
        XCTAssertNil(weatherManager.currentWeather, "Weather data should be nil on network error")
        XCTAssertFalse(weatherManager.isLoading, "Loading should be false after error")
        XCTAssertTrue(weatherManager.errorMessage.contains("Network error"), "Error message should indicate network error")
    }
    
    func test_fetchWeather_timeoutError_handlesAppropriately() async {
        // ARRANGE: Configure timeout scenario
        mockNetworkService.configure(scenario: .timeout)
        
        // ACT: Attempt to fetch weather
        weatherManager.fetchWeather(for: "Sydney")
        
        // Wait for async operation
        try? await Task.sleep(nanoseconds: 100_000_000)
        
        // ASSERT: Timeout should be handled
        XCTAssertNil(weatherManager.currentWeather, "Weather data should be nil on timeout")
        XCTAssertFalse(weatherManager.isLoading, "Loading should be false after timeout")
        XCTAssertTrue(weatherManager.errorMessage.contains("Network error"), "Should show network error message")
    }
    
    // MARK: - Invalid Data Handling Tests
    
    func test_fetchWeather_invalidJSON_handlesParsingError() async {
        // ARRANGE: Invalid JSON that can't be parsed
        let invalidJSON = "{ invalid json }".data(using: .utf8)!
        mockNetworkService.configure(scenario: .success(invalidJSON))
        
        // ACT: Attempt to fetch weather
        weatherManager.fetchWeather(for: "Berlin")
        
        // Wait for async operation
        try? await Task.sleep(nanoseconds: 100_000_000)
        
        // ASSERT: Parsing error should be handled
        XCTAssertNil(weatherManager.currentWeather, "Weather data should be nil when JSON parsing fails")
        XCTAssertFalse(weatherManager.isLoading, "Loading should be false after parsing error")
        XCTAssertTrue(weatherManager.errorMessage.contains("Failed to parse"), "Should show JSON parsing error message")
    }
    
    func test_fetchWeather_emptyResponse_handlesGracefully() async {
        // ARRANGE: Empty data response
        mockNetworkService.configure(scenario: .success(Data()))
        
        // ACT: Attempt to fetch weather  
        weatherManager.fetchWeather(for: "Madrid")
        
        // Wait for async operation
        try? await Task.sleep(nanoseconds: 100_000_000)
        
        // ASSERT: Empty response should be handled
        XCTAssertNil(weatherManager.currentWeather, "Weather data should be nil for empty response")
        XCTAssertFalse(weatherManager.isLoading, "Loading should be false after empty response")
        XCTAssertFalse(weatherManager.errorMessage.isEmpty, "Should show error message for empty response")
    }
    
    // MARK: - Async Behavior Tests
    
    func test_fetchWeather_setsLoadingStateCorrectly() {
        // ARRANGE: Configure slow network response
        mockNetworkService.configure(scenario: .success(Data()))
        mockNetworkService.requestDelay = 0.2 // 200ms delay
        
        // ACT: Start weather fetch
        XCTAssertFalse(weatherManager.isLoading, "Loading should start as false")
        
        weatherManager.fetchWeather(for: "Rome")
        
        // ASSERT: Loading should be set during request
        // Note: In real implementation, isLoading would be set to true at start of fetchWeather
        // This tests the current implementation behavior
    }
    
    func test_fetchWeather_multipleRequests_handlesCorrectly() async {
        // ARRANGE: Configure successful responses
        mockNetworkService.configure(scenario: .success("""
        {
            "name": "TestCity",
            "main": {"temp": 20.0},
            "weather": [{"description": "test"}]
        }
        """.data(using: .utf8)!))
        
        // ACT: Make multiple requests
        weatherManager.fetchWeather(for: "City1")
        weatherManager.fetchWeather(for: "City2")
        weatherManager.fetchWeather(for: "City3")
        
        // Wait for all async operations
        try? await Task.sleep(nanoseconds: 200_000_000) // 0.2 second
        
        // ASSERT: Should have made multiple network requests
        XCTAssertEqual(mockNetworkService.capturedURLs.count, 3, "Should make request for each city")
        XCTAssertTrue(mockNetworkService.capturedURLs[0].absoluteString.contains("City1"))
        XCTAssertTrue(mockNetworkService.capturedURLs[1].absoluteString.contains("City2"))
        XCTAssertTrue(mockNetworkService.capturedURLs[2].absoluteString.contains("City3"))
    }
}

/*
 * REFACTORING BENEFITS DEMONSTRATED:
 *
 * 1. FAST TESTS: No real network calls - tests run in milliseconds instead of seconds
 * 2. RELIABLE TESTS: No dependency on internet connection or external APIs
 * 3. ERROR TESTING: Can simulate network failures, timeouts, malformed responses
 * 4. EDGE CASE TESTING: Can test scenarios difficult to reproduce with real network
 * 5. ASYNC TESTING: Proper async/await testing patterns for modern iOS development
 * 6. ISOLATION: Tests don't interfere with each other or external services
 *
 * BEFORE NETWORK DEPENDENCY BREAKING:
 * - Tests required internet connection and working API
 * - Network failures caused random test failures
 * - Impossible to test error scenarios reliably
 * - Tests were slow (waiting for real network requests)
 * - Tests could hit API rate limits or affect server metrics
 *
 * AFTER NETWORK DEPENDENCY BREAKING:
 * - All network behavior can be tested without real HTTP calls
 * - Error scenarios are easily reproducible and testable
 * - Tests are fast, reliable, and deterministic
 * - Can test edge cases like timeouts, malformed responses, etc.
 * - Tests serve as living documentation of network behavior
 */

File: solutions/RefactoredWeatherKotlin/app/src/test/java/com/example/legacyweatherkotlin/NetworkDependencyTests.kt
package com.example.legacyweatherkotlin

import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test

/**
 * NetworkDependencyTests - Exercise 4: Network Dependency Breaking
 * 
 * PURPOSE: Demonstrate how breaking the network dependency enables fast, reliable tests
 * 
 * BENEFITS OF NETWORK ABSTRACTION:
 * 1. SPEED: No actual HTTP calls, tests run in milliseconds
 * 2. RELIABILITY: No flaky tests due to network issues or API downtime
 * 3. CONTROL: Can simulate any network scenario (success, failure, timeout, errors)
 * 4. ISOLATION: Tests focus on business logic, not network infrastructure
 * 5. DETERMINISM: Same input always produces same output
 * 
 * PATTERN: Dependency Injection with Interface Segregation
 * - WeatherNetworkService interface defines contract
 * - RetrofitWeatherNetworkService implements real HTTP calls
 * - MockNetworkService provides controllable test double
 * 
 * COMPARISON TO SWIFT:
 * - Swift uses async/await pattern
 * - Kotlin uses callback pattern (matches existing Retrofit API)
 * - Both achieve same goal: testable network interactions
 */
class NetworkDependencyTests {
    
    /**
     * MockNetworkService - Test Double Pattern
     * 
     * PURPOSE: Simulates network behavior without actual HTTP calls
     * PATTERN: Test Double (specifically a Configurable Mock)
     * 
     * CAPABILITIES:
     * - Return pre-configured success responses
     * - Simulate various error scenarios
     * - Track call history for verification
     * - Provide immediate (synchronous-like) responses for testing
     */
    private class MockNetworkService : WeatherNetworkService {
        // Configuration: Set response before calling fetchWeather
        var mockResponse: Result<WeatherResponse>? = null
        
        // Tracking: Verify what was requested
        var lastCity: String? = null
        var lastApiKey: String? = null
        var callCount = 0
        
        override fun fetchWeather(
            city: String,
            apiKey: String,
            callback: (Result<WeatherResponse>) -> Unit
        ) {
            // Track the call
            lastCity = city
            lastApiKey = apiKey
            callCount++
            
            // Return configured response
            mockResponse?.let { callback(it) }
                ?: callback(Result.failure(Exception("MockNetworkService not configured")))
        }
        
        // Helper methods for test configuration
        fun setSuccessResponse(response: WeatherResponse) {
            mockResponse = Result.success(response)
        }
        
        fun setFailureResponse(error: Throwable) {
            mockResponse = Result.failure(error)
        }
        
        fun reset() {
            mockResponse = null
            lastCity = null
            lastApiKey = null
            callCount = 0
        }
    }
    
    private lateinit var mockNetworkService: MockNetworkService
    
    @Before
    fun setUp() {
        // PATTERN: Test Isolation - each test starts with clean state
        mockNetworkService = MockNetworkService()
        WeatherSingleton.setNetworkService(mockNetworkService)
        WeatherSingleton.suppressLogging = true // Suppress Android Log calls in unit tests
        WeatherSingleton.clearWeatherData()
    }
    
    @After
    fun tearDown() {
        // PATTERN: Test Cleanup - restore original state
        WeatherSingleton.resetNetworkService()
        WeatherSingleton.suppressLogging = false // Restore logging
        WeatherSingleton.clearWeatherData()
    }
    
    // MARK: - Test Helpers
    
    /**
     * Create a mock WeatherResponse for testing
     * BENEFIT: Centralized test data creation, easy to maintain
     */
    private fun createMockWeatherResponse(
        cityName: String = "London",
        temp: Double = 293.15,
        description: String = "clear sky",
        humidity: Int = 65,
        windSpeed: Double = 5.5,
        icon: String = "01d",
        feelsLike: Double = 291.15,
        pressure: Int = 1013
    ): WeatherResponse {
        return WeatherResponse(
            name = cityName,
            main = Main(
                temp = temp,
                feels_like = feelsLike,
                humidity = humidity,
                pressure = pressure
            ),
            weather = listOf(
                Weather(
                    main = "Clear",
                    description = description,
                    icon = icon
                )
            ),
            wind = Wind(speed = windSpeed)
        )
    }
    
    // MARK: - Basic Network Dependency Tests
    
    @Test
    fun `test network service injection is used`() {
        // BENEFIT: Verify dependency injection works
        val response = createMockWeatherResponse(cityName = "Paris")
        mockNetworkService.setSuccessResponse(response)
        
        WeatherSingleton.fetchWeather("Paris")
        
        // Verify mock was called (proves injection works)
        assertEquals("Paris", mockNetworkService.lastCity)
        assertEquals(1, mockNetworkService.callCount)
    }
    
    @Test
    fun `test successful network call updates weather data`() {
        // BENEFIT: Test business logic without real network
        val response = createMockWeatherResponse(
            cityName = "Tokyo",
            temp = 298.15,
            description = "partly cloudy"
        )
        mockNetworkService.setSuccessResponse(response)
        
        WeatherSingleton.fetchWeather("Tokyo")
        
        // Verify state updates
        val weather = WeatherSingleton.currentWeather.value
        assertNotNull("Weather data should be set", weather)
        assertEquals("Tokyo", weather?.city)
        assertEquals(298.15, weather?.temperature ?: 0.0, 0.01)
        assertEquals("Partly Cloudy", weather?.description) // Capitalized
    }
    
    @Test
    fun `test network failure sets error message`() {
        // BENEFIT: Test error handling without causing real errors
        mockNetworkService.setFailureResponse(
            Exception("Connection timeout")
        )
        
        WeatherSingleton.fetchWeather("Berlin")
        
        // Verify error handling
        val error = WeatherSingleton.errorMessage.value
        assertTrue("Error message should contain failure info", 
            error.contains("Network error") || error.contains("Connection timeout"))
        assertNull("Weather data should remain null on error", 
            WeatherSingleton.currentWeather.value)
    }
    
    @Test
    fun `test loading state management during network call`() {
        // BENEFIT: Test UI state transitions without real network delay
        val response = createMockWeatherResponse()
        mockNetworkService.setSuccessResponse(response)
        
        // Initial state
        assertFalse("Should not be loading initially", 
            WeatherSingleton.isLoading.value)
        
        WeatherSingleton.fetchWeather("London")
        
        // After network call completes (mock responds immediately)
        assertFalse("Should not be loading after completion", 
            WeatherSingleton.isLoading.value)
    }
    
    // MARK: - Error Scenario Tests
    
    @Test
    fun `test HTTP error handling`() {
        // BENEFIT: Simulate specific HTTP errors without server
        mockNetworkService.setFailureResponse(
            Exception("HTTP 404: Not Found")
        )
        
        WeatherSingleton.fetchWeather("InvalidCity")
        
        assertTrue("Error message should mention network error",
            WeatherSingleton.errorMessage.value.contains("Network error"))
        assertNull("Weather should be null on error", 
            WeatherSingleton.currentWeather.value)
    }
    
    @Test
    fun `test network timeout simulation`() {
        // BENEFIT: Test timeout handling without waiting
        mockNetworkService.setFailureResponse(
            Exception("Request timed out")
        )
        
        WeatherSingleton.fetchWeather("RemoteCity")
        
        val error = WeatherSingleton.errorMessage.value
        assertTrue("Should handle timeout error",
            error.contains("Network error") || error.contains("timed out"))
    }
    
    @Test
    fun `test malformed response handling`() {
        // BENEFIT: Test edge cases that are hard to reproduce with real API
        // This simulates what would happen if API returned unexpected data
        mockNetworkService.setFailureResponse(
            Exception("JSON parsing error")
        )
        
        WeatherSingleton.fetchWeather("TestCity")
        
        assertNotNull("Should set error message", 
            WeatherSingleton.errorMessage.value)
        assertNull("Weather should remain null", 
            WeatherSingleton.currentWeather.value)
    }
    
    // MARK: - Multiple Cities Tests
    
    @Test
    fun `test fetching different cities updates data correctly`() {
        // BENEFIT: Test state management across multiple calls
        
        // First city
        val londonResponse = createMockWeatherResponse(
            cityName = "London",
            temp = 288.15
        )
        mockNetworkService.setSuccessResponse(londonResponse)
        WeatherSingleton.fetchWeather("London")
        
        assertEquals("London", WeatherSingleton.currentWeather.value?.city)
        assertEquals(288.15, WeatherSingleton.currentWeather.value?.temperature ?: 0.0, 0.01)
        
        // Second city
        val parisResponse = createMockWeatherResponse(
            cityName = "Paris",
            temp = 290.15
        )
        mockNetworkService.setSuccessResponse(parisResponse)
        WeatherSingleton.fetchWeather("Paris")
        
        assertEquals("Paris", WeatherSingleton.currentWeather.value?.city)
        assertEquals(290.15, WeatherSingleton.currentWeather.value?.temperature ?: 0.0, 0.01)
    }
    
    @Test
    fun `test cached data is not returned for different city`() {
        // BENEFIT: Test cache invalidation logic with controlled time
        
        // Setup time provider for caching test
        val mockTime = MockTimeProvider()
        mockTime.setTime(1000000L)
        WeatherSingleton.setTimeProvider(mockTime)
        
        // Fetch London
        val londonResponse = createMockWeatherResponse(cityName = "London")
        mockNetworkService.setSuccessResponse(londonResponse)
        WeatherSingleton.fetchWeather("London")
        
        assertEquals(1, mockNetworkService.callCount)
        
        // Fetch Paris (should not use London's cache)
        val parisResponse = createMockWeatherResponse(cityName = "Paris")
        mockNetworkService.setSuccessResponse(parisResponse)
        mockTime.advanceTime(1000) // Advance time but still within cache window
        WeatherSingleton.fetchWeather("Paris")
        
        // Should make another network call (different city)
        assertEquals(2, mockNetworkService.callCount)
        assertEquals("Paris", WeatherSingleton.currentWeather.value?.city)
        
        // Cleanup
        WeatherSingleton.setTimeProvider(SystemTimeProvider())
    }
    
    // MARK: - API Key Tests
    
    @Test
    fun `test correct API key is passed to network service`() {
        // BENEFIT: Verify configuration is passed correctly
        val response = createMockWeatherResponse()
        mockNetworkService.setSuccessResponse(response)
        
        WeatherSingleton.fetchWeather("London")
        
        assertNotNull("API key should be passed", mockNetworkService.lastApiKey)
        assertFalse("API key should not be empty", 
            mockNetworkService.lastApiKey.isNullOrEmpty())
    }
    
    // MARK: - Weather Data Transformation Tests
    
    @Test
    fun `test weather response is transformed correctly`() {
        // BENEFIT: Test data transformation without network
        val response = createMockWeatherResponse(
            cityName = "Sydney",
            temp = 295.15,
            description = "scattered clouds",
            humidity = 70,
            windSpeed = 8.2,
            icon = "03d",
            feelsLike = 293.15,
            pressure = 1015
        )
        mockNetworkService.setSuccessResponse(response)
        
        WeatherSingleton.fetchWeather("Sydney")
        
        val weather = WeatherSingleton.currentWeather.value
        assertNotNull(weather)
        assertEquals("Sydney", weather?.city)
        assertEquals(295.15, weather?.temperature ?: 0.0, 0.01)
        assertEquals("Scattered Clouds", weather?.description)
        assertEquals(70, weather?.humidity)
        assertEquals(8.2, weather?.windSpeed ?: 0.0, 0.01)
        assertEquals("03d", weather?.icon)
        assertEquals(293.15, weather?.feelsLike ?: 0.0, 0.01)
        assertEquals(1015, weather?.pressure)
    }
    
    @Test
    fun `test description capitalization in transformation`() {
        // BENEFIT: Test string transformation logic in isolation
        val response = createMockWeatherResponse(
            description = "light rain shower"
        )
        mockNetworkService.setSuccessResponse(response)
        
        WeatherSingleton.fetchWeather("London")
        
        assertEquals("Light Rain Shower", 
            WeatherSingleton.currentWeather.value?.description)
    }
    
    // MARK: - Extreme Weather Data Tests
    
    @Test
    fun `test extreme temperature values`() {
        // BENEFIT: Test edge cases that are hard to get from real API
        val response = createMockWeatherResponse(
            temp = 233.15, // -40°C/-40°F
            feelsLike = 228.15 // -45°C
        )
        mockNetworkService.setSuccessResponse(response)
        
        WeatherSingleton.fetchWeather("Antarctica")
        
        val weather = WeatherSingleton.currentWeather.value
        assertNotNull(weather)
        assertEquals(233.15, weather?.temperature ?: 0.0, 0.01)
        assertEquals(228.15, weather?.feelsLike ?: 0.0, 0.01)
    }
    
    @Test
    fun `test very high temperature values`() {
        // BENEFIT: Test extreme high temperatures
        val response = createMockWeatherResponse(
            temp = 323.15, // 50°C/122°F
            feelsLike = 328.15 // 55°C
        )
        mockNetworkService.setSuccessResponse(response)
        
        WeatherSingleton.fetchWeather("Death Valley")
        
        val weather = WeatherSingleton.currentWeather.value
        assertNotNull(weather)
        assertEquals(323.15, weather?.temperature ?: 0.0, 0.01)
    }
    
    @Test
    fun `test zero humidity`() {
        // BENEFIT: Test boundary conditions
        val response = createMockWeatherResponse(humidity = 0)
        mockNetworkService.setSuccessResponse(response)
        
        WeatherSingleton.fetchWeather("Desert")
        
        assertEquals(0, WeatherSingleton.currentWeather.value?.humidity)
    }
    
    @Test
    fun `test maximum humidity`() {
        // BENEFIT: Test boundary conditions
        val response = createMockWeatherResponse(humidity = 100)
        mockNetworkService.setSuccessResponse(response)
        
        WeatherSingleton.fetchWeather("Rainforest")
        
        assertEquals(100, WeatherSingleton.currentWeather.value?.humidity)
    }
    
    @Test
    fun `test calm wind conditions`() {
        // BENEFIT: Test zero/minimal values
        val response = createMockWeatherResponse(windSpeed = 0.0)
        mockNetworkService.setSuccessResponse(response)
        
        WeatherSingleton.fetchWeather("Calm Area")
        
        assertEquals(0.0, WeatherSingleton.currentWeather.value?.windSpeed ?: -1.0, 0.01)
    }
    
    @Test
    fun `test hurricane force winds`() {
        // BENEFIT: Test extreme values
        val response = createMockWeatherResponse(windSpeed = 50.0)
        mockNetworkService.setSuccessResponse(response)
        
        WeatherSingleton.fetchWeather("Hurricane Zone")
        
        assertEquals(50.0, WeatherSingleton.currentWeather.value?.windSpeed ?: 0.0, 0.01)
    }
    
    // MARK: - Cache Interaction with Network Tests
    
    @Test
    fun `test cache prevents unnecessary network calls`() {
        // BENEFIT: Test caching behavior with controllable time and network
        val mockTime = MockTimeProvider()
        mockTime.setTime(1000000L)
        WeatherSingleton.setTimeProvider(mockTime)
        
        val response = createMockWeatherResponse()
        mockNetworkService.setSuccessResponse(response)
        
        // First call - should hit network
        WeatherSingleton.fetchWeather("London")
        assertEquals(1, mockNetworkService.callCount)
        
        // Second call within cache window - should use cache
        mockTime.advanceTime(60000) // 1 minute later
        WeatherSingleton.fetchWeather("London")
        assertEquals("Should not make second network call (cache hit)", 
            1, mockNetworkService.callCount)
        
        // Cleanup
        WeatherSingleton.setTimeProvider(SystemTimeProvider())
    }
    
    @Test
    fun `test cache expires and makes new network call`() {
        // BENEFIT: Test cache expiration with controlled time
        val mockTime = MockTimeProvider()
        mockTime.setTime(1000000L)
        WeatherSingleton.setTimeProvider(mockTime)
        
        val response = createMockWeatherResponse()
        mockNetworkService.setSuccessResponse(response)
        
        // First call
        WeatherSingleton.fetchWeather("London")
        assertEquals(1, mockNetworkService.callCount)
        
        // Call after cache expiry
        mockTime.advanceTime(400000) // 6+ minutes later
        WeatherSingleton.fetchWeather("London")
        assertEquals("Should make new network call (cache expired)", 
            2, mockNetworkService.callCount)
        
        // Cleanup
        WeatherSingleton.setTimeProvider(SystemTimeProvider())
    }
    
    // MARK: - Error Recovery Tests
    
    @Test
    fun `test recovery after network error`() {
        // BENEFIT: Test error recovery without real network issues
        
        // First call fails
        mockNetworkService.setFailureResponse(Exception("Network error"))
        WeatherSingleton.fetchWeather("London")
        
        assertNull("Weather should be null after error", 
            WeatherSingleton.currentWeather.value)
        assertNotEquals("", WeatherSingleton.errorMessage.value)
        
        // Second call succeeds
        val response = createMockWeatherResponse()
        mockNetworkService.setSuccessResponse(response)
        WeatherSingleton.fetchWeather("London")
        
        assertNotNull("Weather should be set after recovery", 
            WeatherSingleton.currentWeather.value)
        assertEquals("Error should be cleared", "", WeatherSingleton.errorMessage.value)
    }
    
    // MARK: - MockTimeProvider (reused from TimeProviderTests)
    
    private class MockTimeProvider : TimeProvider {
        private var currentTimeMillis: Long = 0
        
        fun setTime(timeMillis: Long) {
            currentTimeMillis = timeMillis
        }
        
        fun advanceTime(millis: Long) {
            currentTimeMillis += millis
        }
        
        override fun currentTimeMillis(): Long = currentTimeMillis
    }
}
